### PR TITLE
Add a phone number field to the form (and add back end support)

### PR DESCRIPTION
**Pull request by AWare**
This depends on: https://github.com/guardian/membership-common/pull/191
and: https://github.com/guardian/identity-play-auth/pull/10

### CODE REVIEW

File: app/views/fragments/checkout/fieldsPersonal.scala.html [hunk 1]
```diff
@@ -37,6 +38,17 @@
     }
 </div>
 
+    @* ===== Phone ===== *@
+    @if(touchpointBackendResolution.validTestUserCredentialOpt){
```
Comment (joelochlann) [inline app/views/fragments/checkout/fieldsPersonal.scala.html]: indentation

File: app/services/IdentityService.scala [hunk 2]
```diff
@@ -201,12 +210,13 @@ object IdentityApiClient extends IdentityApiClient with LazyLogging {
...
 
-  override val updateUserDetails: (PersonalData, UserId, AccessCredentials.Cookies) => Future[WSResponse] = { (user, userId, authCookies) =>
+  override val updateUserDetails: (PersonalData, UserId, AccessCredentials.Cookies) => Future[WSResponse] = { (personalData, userId, authCookies) =>
     val endpoint = authoriseCall(WS.url(s"$identityEndpoint/user/${userId.id}"))
-    val userJson: JsObject = user
+    val userJson: JsObject = personalData
     val updatedFields =
       createOnlyFields.foldLeft(userJson) { (map, field) => map - field }
 
+    println(userJson)
```
Comment (joelochlann) [inline app/services/IdentityService.scala]: whoops, assume this was for debug purposes?

File: app/model/SubscriptionData.scala [hunk 3]
```diff
@@ -71,15 +74,17 @@ object SubscriptionData {
...
+      address = u.address,
+      telephoneNumber = u.privateFields.flatMap(_.telephoneNumber).flatMap(_.localNumber)
     )
 
     val blankPaymentData = DirectDebitData("", "", "")
 
     val blankRatePlanId = ProductRatePlanId("") // this makes me very nervous indeed
 
-    SubscriptionData(personalData, blankPaymentData, blankRatePlanId, promoCode)
+    SubscriptionData(personalData, blankPaymentData, blankRatePlanId ,promoCode)
```
Comment (joelochlann) [inline app/model/SubscriptionData.scala]: why was this line changed?

File: app/views/fragments/checkout/fieldsPersonal.scala.html [hunk 4]
```diff
@@ -37,6 +38,17 @@
...
+<div class="form-field js-checkout-phone-number">
+    <label class="label" for="telephoneNumber">Phone Number</label>
```
Comment (joelochlann) [inline app/views/fragments/checkout/fieldsPersonal.scala.html]: `for` should reference the id of the input, not the name
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label

File: app/services/IdentityService.scala [hunk 5]
```diff
@@ -124,6 +132,7 @@ object IdentityApiClient extends IdentityApiClient with LazyLogging {
   implicit class FutureWSLike(eventualResponse: Future[WSResponse]) {
     /**
      * Example of ID API Response: `{"status":"error","errors":[{"message":"Forbidden","description":"Field access denied","context":"privateFields.billingAddress1"}]}`
+ *
```
Comment (joelochlann) [inline app/services/IdentityService.scala]: Not sure why this line is here?

File: app/services/IdentityService.scala [hunk 6]
```diff
@@ -88,8 +90,14 @@ object PersonalDataJsonSerialiser {
         "billingAddress3" -> personalData.address.town,
         "billingAddress4" -> personalData.address.countyOrState,
         "billingPostcode" -> personalData.address.postCode,
-        "billingCountry" -> personalData.country.name
-      ),
+        "billingCountry"  -> personalData.country.name
+      ).++(
+         telephoneNumber.map {t =>
+        Json.obj(
```
Comment (joelochlann) [inline app/services/IdentityService.scala]: The indentation makes this particularly hard to read. Also, `fold` might be more idiomatic here

File: app/model/SubscriptionData.scala [hunk 7]
```diff
@@ -39,7 +38,9 @@ case class PersonalData(first: String,
                         last: String,
                         email: String,
                         receiveGnmMarketing: Boolean,
-                        address: Address
+                        address: Address,
+                        telephoneNumber: Option[String] = None
+
```
Comment (joelochlann) [inline app/model/SubscriptionData.scala]: rogue newline

**Comment by joelochlann** [comment]
sorry, missed the [WIP] :stuck_out_tongue_winking_eye:

**Comment by AWare** [comment]
Ready for code review, need to confirm access to identity api phone-number fields before release.

File: app/services/IdentityService.scala [hunk 8]
```diff
@@ -201,9 +207,9 @@ object IdentityApiClient extends IdentityApiClient with LazyLogging {
       .withCloudwatchMonitoringOfPut
   }
 
-  override val updateUserDetails: (PersonalData, UserId, AccessCredentials.Cookies) => Future[WSResponse] = { (user, userId, authCookies) =>
-    val endpoint = authoriseCall(WS.url(s"$identityEndpoint/user/${userId.id}"))
-    val userJson: JsObject = user
+  override val updateUserDetails: (PersonalData, AccessCredentials.Cookies) => Future[WSResponse] = { (personalData, authCookies) =>
+    val endpoint = authoriseCall(WS.url(s"$identityEndpoint/user/me"))
```
Comment (tomverran) [inline app/services/IdentityService.scala]: We/re changing the identity endpoint?!

Comment (tomverran) [inline app/services/IdentityService.scala]: This was determined to be okay

**Comment by joelochlann** [comment]
:+1: but can you please add before/after screenshots to show what's changed on the frontend, and provide a bit more context in the initial comment as to why we're doing this?

**Comment by AWare** [comment]
image
image

**Comment by tomverran** [comment]
:+1:
